datasource/src/main/scala/quasar/physical/mongo/Tunnel.scala

```diff
@@ -0,0 +1,134 @@
+
+import com.jcraft.jsch._
+import com.mongodb.connection.ClusterSettings
+import com.mongodb.{ConnectionString, ServerAddress}
+
+import quasar.Disposable
+
+import scala.collection.JavaConverters._
+
+object Tunnel {
+  import TunnelConfig._, Pass._
+
+  val SessionName: String = "default"
+
+  def apply[F[_]: Sync: ContextShift](config: MongoConfig): F[Disposable[F, ClusterSettings]] =
+    connectionString[F](config.connectionString) flatMap { (conn: ConnectionString) =>
+      val settings: ClusterSettings = ClusterSettings.builder.applyConnectionString(conn).build
+      config.tunnelConfig match {
+        case None =>
+          Disposable(settings, Sync[F].unit).pure[F]
+        case Some(tcfg) =>
+          viaTunnel(tcfg, settings)
+      }
+    }
+
+  def connectionString[F[_]: Sync](s: String): F[ConnectionString] = Sync[F].delay {
+    new ConnectionString(s)
+  }
+
+  def mkJSch[F[_]: Sync]: F[JSch] = Sync[F].delay { new JSch() }
+
+  @SuppressWarnings(Array("org.wartremover.warts.Null"))
+  def mkSession[F[_]: Sync](jsch: JSch, cfg: TunnelConfig): F[Session] = cfg.pass match {
+    case None =>
+      Sync[F].delay { jsch.getSession(cfg.user, cfg.host, cfg.port) }
+    case Some(cred) => cred match {
+      case Password(password) => for {
+        s <- Sync[F].delay { jsch.getSession(cfg.user, cfg.host, cfg.port) }
+        _ <- Sync[F].delay { s.setPassword(password) }
+      } yield s
+      case Identity(prv, passphrase) => for {
+        _ <- Sync[F].delay { jsch.addIdentity(SessionName, prv.getBytes("UTF-8"), null, passphrase map (_.getBytes("UTF-8")) getOrElse null) }
+        s <- Sync[F].delay { jsch.getSession(cfg.user, cfg.host, cfg.port) }
+      } yield s
+    }
+  }
+
+  def userInfo(cfg: TunnelConfig): UserInfo = new UserInfo {
+    override def getPassword(): String = cfg.pass match {
+      case None => ""
+      case Some(pass) => pass match {
+        case Identity(_, _) => ""
+        case Password(p) => p
+      }
+    }
+    override def getPassphrase(): String = cfg.pass match {
+      case None => ""
+      case Some(pass) => pass match {
+        case Identity(_, p) => p getOrElse ""
+        case Password(_) => ""
+      }
+    }
+    override def promptYesNo(s: String): Boolean = true
+    override def promptPassphrase(s: String): Boolean = true
+    override def promptPassword(s: String): Boolean = true
+    override def showMessage(s: String): Unit = ()
+  }
+
+  def setUserInfo[F[_]: Sync](s: Session, u: UserInfo): F[Unit] = Sync[F].delay {
+    s.setUserInfo(u)
+  }
+
+  def serverAddress[F[_]: Sync](settings: ClusterSettings): F[ServerAddress] = {
+    val hosts = settings.getHosts
+    if (hosts.size < 1) Sync[F].raiseError(new Throwable("Server addresses were not provided"))
+    else Sync[F].delay { hosts.get(0) }
+  }
+
+  final case class Tunnel(port: Int)
+
+  def openTunnel[F[_]: Sync](session: Session, address: ServerAddress): F[Tunnel] = Sync[F].delay {
+    session.connect
```

Comment:
Does this need to be done on a blocking thread pool? How are the threading resources handled by JSCH?

Also, random Scala idiom: usually this would be done as `session.connect()`, since `connect` is a side-effecting nullary function. We obviously don't see this much since we simply don't side-effect, but it's nice to keep the idiom in place when we *do* have side-effects (such as here).

Reply:
JSCH creates new thread on `connect` and remove it on `disconnect`. 
It also uses `synchronized` a lot, I'll add `evalOn` to `openTunnel`, thank you!